service/history/shard/controller_impl.go

```diff
@@ -58,33 +49,19 @@ const (
 
 type (
 	ControllerImpl struct {
+		d shardControllerDeps
```

Comment:
why composition here and "inheritance" in shard/context_impl?

Reply:
The reasoning is, direct embedding ("inheritance") is preferable since we don't need the extra field, but it's impossible here so we fall back to composition